### PR TITLE
Update Docker files

### LINKED ISSUE #425: Install libsodium-dev in Dockerfile
Since PR #404, the sodium library is used which requires `libsodium-dev` to be installed.
It should be added to the `Dockerfile`.

### LINKED ISSUE #457: Install curl in Dockerfile
For deployment, it would be convenient to have curl installed in the `trestletech/plumber` docker image. I am talking about the actual CLI tool, not just the library. 

It is commonly used to automate health checks on deployed containers (for example `CMD-SHELL, curl -f localhost:8000/mean || exit 1`).

It takes less than 500KB:

```
The following additional packages will be installed:
  libcurl4
The following NEW packages will be installed:
  curl
The following packages will be upgraded:
  libcurl4
1 upgraded, 1 newly installed, 0 to remove and 55 not upgraded.
Need to get 596 kB of archives.
After this operation, 410 kB of additional disk space will be used.
```

### LINKED ISSUE #459: Cannot install Debian packages in trestletech/plumber
~Installing curl~ Running `apt-get update` in a Dockerfile based on trestletech/plumber fails with the following error:
```
E: Repository 'http://cdn-fastly.deb.debian.org/debian testing InRelease' changed its 'Codename' value from 'buster' to 'bullseye'
The command '/bin/sh -c apt-get update && apt-get install -t unstable -y curl' returned a non-zero code: 10
```

Maybe a rebuild with a more recent version of debian/testing fixes the issue. However, I think the deeper issue is that a Docker image based on debian/testing will always be prone to instabilities. It would be great if we could have a trestletech/plumber image based on rocker/r-ver.

### Example application *or* steps to reproduce the problem

Dockerfile:
```
FROM trestletech/plumber
RUN apt-get update \
&& apt-get install -t unstable -y curl
RUN apt-get install -y libxml2-dev
COPY . /api
EXPOSE 8000
ENTRYPOINT ["R", "-e", "pr <- plumber::plumb(commandArgs()[4]); pr$run(host='0.0.0.0', port=8000, swagger=TRUE)"]
CMD ["/api/plumber.R"]
```

### Describe the problem in detail

```
$ docker build .
Step 1/8 : FROM trestletech/plumber
 ---> f9aa6e6553fb
Step 2/8 : RUN apt-get update && apt-get install -t unstable -y curl
 ---> Running in f8d9f02f20ac
Get:1 http://cdn-fastly.deb.debian.org/debian testing InRelease [117 kB]
Get:2 http://cdn-fastly.deb.debian.org/debian sid InRelease [149 kB]
Get:3 http://cdn-fastly.deb.debian.org/debian sid/main amd64 Packages.diff/Index [27.9 kB]
Ign:3 http://cdn-fastly.deb.debian.org/debian sid/main amd64 Packages.diff/Index
Get:4 http://cdn-fastly.deb.debian.org/debian sid/main amd64 Packages [8,389 kB]
Reading package lists...
E: Repository 'http://cdn-fastly.deb.debian.org/debian testing InRelease' changed its 'Codename' value from 'buster' to 'bullseye'
The command '/bin/sh -c apt-get update && apt-get install -t unstable -y curl' returned a non-zero code: 100
```

### LINKED ISSUE #511: commandArgs()[4] failing for plumber Docker launch: "File does not exist -e"
A plumber Dockerfile was created that added extra arguments meaning the "boilerplate" `commandArgs()[4]` didn't work for launching the server.

This was discovered after adding some exploratory functions as seen in the screenshots below

image

image

### Describe the problem in detail

In all my plumber Dockerfiles to date I have launched the server via:

```r
pr <- plumber::plumb(commandArgs()[4]); pr$host(host='0.0.0.0', port=as.numeric(sys.getenv('PORT'))
```

This relies on R being executed via a startup similar to:

```R -e "pr <- plumber::plumb(commandArgs()[4]); pr$host(host='0.0.0.0', port=as.numeric(sys.getenv('PORT'))" api.R```

However in the screenshot above it can be seen `--no-save` and `--no-restore` was inserted, so breaking the positional argument of `commandArgs()` which needed to be `commandArgs()[6]` rather than `commandArgs()[4]` (pointing at the api.R file)

```R --no-save --no-restore -e "pr <- plumber::plumb(commandArgs()[6]); pr$host(host='0.0.0.0', port=as.numeric(sys.getenv('PORT'))" api.R```

My questions are :)

* Would anyone know why this happened when launching R in this Docker container in particular? Can it be suppressed?
* Is it possible to use something more robust than positional arguments in parsing out the command arguments?

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #511
Fixes #425
Fixes #457
Fixes #459 

* Adds Docker images to https://hub.docker.com/r/rstudio/plumber
* Builds tags for
  * `v0.4.6`, `latest` for the plumber@v0.4.6
  * (As new releases are made, all _stable_ releases should be maintained for a reasonable amount of time.)
  * `next` for master branch
* Reverses command line arguments to retrieve file which avoids issue when extra arguments are provided

### CODE REVIEW

**Comment by schloerke** [comment]
@mskyttner and @MarkEdmondson1234 . Would these updates work for you? Would you like to see any more changes?

**Comment by MarkEdmondson1234** [comment]
Looks great thanks!

File: Docker/GitHub/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,58 @@
+
+ARG R_VERSION=latest
+
+FROM rocker/r-ver:${R_VERSION}
+
+ARG PLUMBER_BRANCH=master
```
Comment (cpsievert) [inline Docker/GitHub/Dockerfile:6]:
```suggestion
ARG PLUMBER_REF=master
```

File: Docker/GitHub/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  curl \
+  libsodium-dev \
+  libxml2-dev
+
+RUN install2.r remotes
+
+# But cache at this step
+# https://stackoverflow.com/a/55621942/591574
+ADD https://github.com/rstudio/plumber/commits/ _docker_cache
+RUN Rscript -e "remotes::install_github('rstudio/plumber@${PLUMBER_BRANCH}')"
```
Comment (cpsievert) [inline Docker/GitHub/Dockerfile:22]:
```suggestion
RUN Rscript -e "remotes::install_github('rstudio/plumber@${PLUMBER_REF}')"
```

File: Docker/CRAN/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+# Extend the rstudio/plumber:TAG Dockerfile / build your own custom image adding debian packages and your own api:
+
+## ./Dockerfile
+#   FROM rstudio/plumber:v0.4.6
+#   RUN apt-get update -qq && apt-get install -y \
+#     [list-your-debian-packages-here]
+#   # add app files from host's present working dir
+#   COPY . /api
+#   # set default startup command to run the app's "plumber.R" file
+#   CMD ["/api/plumber.R"]
```
Comment (cpsievert) [inline Docker/CRAN/Dockerfile:55]: Get rid of this file and just use the other

File: .github/workflows/docker.yaml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          workdir: Docker/CRAN
+          # always overwrite the latest version with the CRAN version
+          tags: "v0.4.6,latest"
+          extra_buildargs: ",PLUMBER_VERSION=0.4.6"
+
+        - name: GitHub
+          workdir: Docker/GitHub
+          # 'next' tag signifies the next release
+          tags: "next"
+          extra_buildargs: ",PLUMBER_BRANCH=master"
```
Comment (cpsievert) [inline .github/workflows/docker.yaml:34]:
```suggestion
          extra_buildargs: ",PLUMBER_REF=master"
```

File: .github/workflows/docker.yaml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      matrix:
+        r_version:
+          - "latest"
+
+        config:
+        - name: CRAN
+          workdir: Docker/CRAN
+          # always overwrite the latest version with the CRAN version
+          tags: "v0.4.6,latest"
+          extra_buildargs: ",PLUMBER_VERSION=0.4.6"
```
Comment (cpsievert) [inline .github/workflows/docker.yaml:28]:
```suggestion
        - name: CRAN
          # always overwrite the latest version with the CRAN version
          tags: "v0.4.6,latest"
          extra_buildargs: ",PLUMBER_REF=v0.4.6"
```

File: Dockerfile [hunk 6]
```diff
@@ -1,16 +1,60 @@
...
+  curl \
+  libsodium-dev \
+  libxml2-dev
+
+RUN install2.r remotes
 
-## RUN R -e 'install.packages(c("remotes"))'
-## RUN R -e 'remotes::install_github("rstudio/plumber")'
-RUN install2.r plumber
+# But cache at this step
```
Comment (schloerke) [inline Dockerfile:20]: spelling 

```suggestion
# Bust cache at this step
```

File: Dockerfile [hunk 7]
```diff
@@ -1,16 +1,60 @@
...
+  libxml2-dev
+
+RUN install2.r remotes
 
-## RUN R -e 'install.packages(c("remotes"))'
-## RUN R -e 'remotes::install_github("rstudio/plumber")'
-RUN install2.r plumber
+# Bust cache at this step
+# https://stackoverflow.com/a/55621942/591574
+ADD https://github.com/rstudio/plumber/commits/ _docker_cache
```
Comment (schloerke) [inline Dockerfile:22]:
```suggestion
## Remove this comment to always bust the Docker cache at this step
## https://stackoverflow.com/a/55621942/591574
#ADD https://github.com/rstudio/plumber/commits/ _docker_cache
```

**Pull request merged**